include/net/websocket.h

```diff
@@ -0,0 +1,136 @@
+#define WS_FLAG_CLOSE  0x00000008
+#define WS_FLAG_PING   0x00000010
+#define WS_FLAG_PONG   0x00000011
+
+enum ws_opcode  {
+	WS_OPCODE_CONTINUE     = 0x00,
+	WS_OPCODE_DATA_TEXT    = 0x01,
+	WS_OPCODE_DATA_BINARY  = 0x02,
+	WS_OPCODE_CLOSE        = 0x08,
+	WS_OPCODE_PING         = 0x09,
+	WS_OPCODE_PONG         = 0x0A,
+};
+
+/**
+ * @brief Send websocket msg to peer.
+ *
+ * @details The function will automatically add websocket header to the
+ * message.
+ *
+ * @param ctx Websocket context.
+ * @param payload Websocket data to send.
+ * @param payload_len Length of the data to be sent.
+ * @param opcode Operation code (text, binary, ping, pong, close)
+ * @param mask Mask the data, see RFC 6455 for details
+ * @param final Is this final message for this message send. If final == false,
+ * then the first message must have valid opcode and subsequent messages must
+ * have opcode WS_OPCODE_CONTINUE. If final == true and this is the only
+ * message, then opcode should have proper opcode (text or binary) set.
+ * @param user_send_data User specific data to this connection. This is passed
+ * as a parameter to sent cb after the packet has been sent.
+ *
+ * @return 0 if ok, <0 if error.
+ */
+int ws_send_msg(struct http_ctx *ctx, u8_t *payload, size_t payload_len,
+		enum ws_opcode opcode, bool mask, bool final,
+		void *user_send_data);
+
+/**
+ * @brief Send message to client.
+ *
+ * @details The function will automatically add websocket header to the
+ * message.
+ *
+ * @param ctx Websocket context.
+ * @param payload Websocket data to send.
+ * @param payload_len Length of the data to be sent.
+ * @param opcode Operation code (text, binary, ping ,pong ,close)
+ * @param final Is this final message for this data send
+ * @param user_send_data User specific data to this connection. This is passed
+ * as a parameter to sent cb after the packet has been sent.
+ *
+ * @return 0 if ok, <0 if error.
+ */
+static inline int ws_send_msg_to_client(struct http_ctx *ctx,
+					u8_t *payload,
+					size_t payload_len,
+					enum ws_opcode opcode,
+					bool final,
+					void *user_send_data)
+{
+	return ws_send_msg(ctx, payload, payload_len, opcode, false, final,
+			   user_send_data);
+}
+
+/**
+ * @brief Strip websocket header from the packet.
+ *
+ * @details The function will remove websocket header from the network packet.
+ *
+ * @param pkt Network packet to send
+ * @param masked The mask status of the message is returned.
+ * @param mask_value The mask value of the message is returned.
+ *
+ * @return Websocket flag value is returned.
+ */
+u32_t ws_strip_header(struct net_pkt *pkt, bool *masked,
+		      u32_t *mask_value);
+
+/**
+ * @brief Mask or unmask a websocket message if needed
+ *
+ * @details The function will either add or remove the masking from the data.
+ *
+ * @param pkt Network packet to process
+ * @param masking_value The mask value to use.
+ */
+void ws_mask_pkt(struct net_pkt *pkt, u32_t masking_value);
```

Comment:
Looking just at the header, it's strange to have such a function as part of the public API. More specifically, I don't think it can work at all: masking_value is actually sequence on 4 bytes, applied with wrap-around to a stream of bytes. Thus, it needs to maintain an offset in 4-byte seq, or it won't be able to apply xform correctly to e.g. pkt of 5 bytes followed by pkt of 1 byte.

Reply:
The function works just fine in my tests, it properly loops over masking_value. Why do you say it cannot work, have you tried it and it failed?

Reply:
No, it's based on the speculative review and knowledge of websocket protocol/experience with implementing it. There's always a chance of reviewer's error.

But what I see in https://github.com/zephyrproject-rtos/zephyr/pull/1271/commits/273e1130af5d38c0551f9906ab8ed1bcf0eb24e8#diff-4b79f0bbb3bc74c2d5bbeb4209f6490eR664 is a typical layering mis-assumption. Websocket records is a stream of bytes, wrapped in TCP stream of bytes, wrapped in IP packets. You receive an IP packet and think that it represents a complete Websocket message. It does not. 1st packet you receive contains 1st byte of Websocket header, 2nd byte - 2nd, etc., one packet per each byte. *That*'s the picture someone working with a streaming protocol like TCP should have in mind, not something else. So, you can't even "read" a mask value from a pkt - you should buffer a stream of bytes of enough length to hold an entire Websocket header, then when you fill it in, you should parse it. Then next packets won't even contain WS header, they will contain WS data (with arbitrary unmasking offset as was pointed out). Then finally one of the next headers will contain 1000 bytes of data followed by next WS header (in the same pkt).

(Yeah, when you implement the code infra to support correct handling of a stream protocol, you'll end up duplicating a lot of BSD Sockets functionality, that point was already shared.)

Reply:
Yes, I see your point. I did not understand your first comment.